### PR TITLE
adding c file of fenwick tree

### LINKED ISSUE #2173: Fenwick Tree
- [x] C
- [x] C++
- [ ] C# - @ritish099
- [ ] CoffeeScript
- [x] Dart - @Rukmini-Meda
- [ ] Go
- [x] Java
- [x] JS
- [x] Kotlin - @JasbirCodeSpace
- [x] Php - @goyalmuskan
- [x] Python
- [ ] Readme - @AJAYSHARMA2012
- [ ] Ruby
- [ ] Rust
- [ ] TypeScript

### PR DESCRIPTION
**Pull request by santushtisharma10**
#2173 Added C implementation of Fenwick Tree Algorithm(Issue #2173 )
Fixes #2173 

#### Checklist:

- [x]  4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- added C file for fenwick tree algorithm

#### Languages Used:
- C
 

#### Files Added:
- fenwick_tree.c
 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by santushtisharma10** [comment]
please review my pr

**Comment by santushtisharma10** [comment]
@jainaman224 please review my pr

File: Fenwick_Tree/fenwick_tree.c [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+
+/*
+   Returns sum of arr[0..index]. This function assumes 
+   that the array is preprocessed and partial sums of 
+   array elements are stored in BITree[]. 
+*/
+int getSum(int BITree[], int index) 
+{ 
+	int sum = 0; // Iniialize result
```
Comment (raghav-dalmia) [inline Fenwick_Tree/fenwick_tree.c]: use 4 space indentation

File: Fenwick_Tree/fenwick_tree.c [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    return getSum(BITree,r)-getSum(BITree,l);
+}
+
+
+// Driver program to test above functions 
+
+int main() 
+{   
+	int n;
+
```
Comment (raghav-dalmia) [inline Fenwick_Tree/fenwick_tree.c]: remove this blank line

File: Fenwick_Tree/fenwick_tree.c [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	
+	printf("%d\n",rangeSum(2,7,BITree));
+	
+	
+	
+	return 0; 
+} 
+
+/*  
+      __INPUT___
```
Comment (raghav-dalmia) [inline Fenwick_Tree/fenwick_tree.c]: This doesn't look good. Left-aligned them.

File: Fenwick_Tree/fenwick_tree.c [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	//Update operation
+
+	freq[4]+=16;
+	updateBIT(BITree,n,4,16);
+	
+	//  Print sum of freq[2....7] after update
+    
+	
+	printf("%d\n",rangeSum(2,7,BITree));
+
```
Comment (raghav-dalmia) [inline Fenwick_Tree/fenwick_tree.c]: remove these extra lines

File: Fenwick_Tree/fenwick_tree.c [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	
+	for(int i=0;i<n;i++)
+	{
+		scanf("%d",&freq[i]);
+	}
+	
+	
+	int *BITree = constructBITree(freq, n); 
+
+ // Print sum of freq[0...5]
```
Comment (raghav-dalmia) [inline Fenwick_Tree/fenwick_tree.c]: improper indentation

File: Fenwick_Tree/fenwick_tree.c [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+int main() 
+{   
+	int n;
+ 
+	scanf("%d",&n);
+	
+	int freq[n];
+	
+	for(int i=0;i<n;i++)
```
Comment (raghav-dalmia) [inline Fenwick_Tree/fenwick_tree.c]: add space after `;` operator

File: Fenwick_Tree/fenwick_tree.c [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+		updateBIT(BITree, n, i, arr[i]); 
+
+	// Uncomment below lines to see contents of BITree[] 
+	//for (int i=1; i<=n; i++) 
+	//	 cout << BITree[i] << " "; 
+
+	return BITree; 
+} 
+
+int rangeSum(int l,int r,int BITree[])
```
Comment (raghav-dalmia) [inline Fenwick_Tree/fenwick_tree.c]: add space after ,

File: Fenwick_Tree/fenwick_tree.c [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+   Fenwick Tree[also known as binary indexed tree] is used when there is a problem of range sum query wiht update i.e.
+   RMQ.Suppose you have an array and you have been asked to find sum in range.It can be done in linear time with
+   static array method.It will be difficult for on to do it in linear time when you have point updates.In this, update 
+   operation incrementing an index by some value.Brute force approach will take O(n^2) time but Fenwick tree will take 
+   O(nlogn) time.
+
+*/
+#include<stdio.h>
+
+/*		 n --> No. of elements present in input array.
```
Comment (raghav-dalmia) [inline Fenwick_Tree/fenwick_tree.c]: improper indentation

**Comment by santushtisharma10** [comment]
@raghav-dalmia sir,i have made the changes as you said,please review them

**Comment by santushtisharma10** [comment]
@raghav-dalmia ,i have made changes as you said,please review them

**Comment by santushtisharma10** [comment]
@raghav-dalmia ,please review my PR

**Comment by raghav-dalmia** [comment]
Some changes are still left. Make sure to resolve them. Also, squash your commits after that.

**Comment by santushtisharma10** [comment]
@raghav-dalmia  please review my pr i have resolved all the issues and squashed into 1 commit as well

**Comment by santushtisharma10** [comment]
@jainaman224 @somya-kapoor @raghav-dalmia please review my pr and tell if any changes has to be made

File: Fenwick_Tree/fenwick_tree.c [hunk 9]
```diff
@@ -0,0 +1,104 @@
+/*
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:1]: File name should be Fenwick_Tree/Fenwick_Tree.c

File: Fenwick_Tree/fenwick_tree.c [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    Fenwick Tree[also known as binary indexed tree] is used when there is a problem of range sum query wiht update i.e.RMQ.
+    Suppose you have an array and you have been asked to find sum in range.It can be done in linear time with static array
+    method.It will be difficult for on to do it in linear time when you have point updates.In this, update operation
+    incrementing an index by some value.Brute force approach will take O(n^2) time but Fenwick tree will take O(nlogn) time.
+*/
+#include <stdio.h>
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:7]: Leave a line before it

File: Fenwick_Tree/fenwick_tree.c [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+/*  n --> No. of elements present in input array. 
+    BITree[0..n] --> Array that represents Fenwick Tree(Binary Indexed Tree). 
+    arr[0..n-1] --> Input array for which prefix sum is evaluated.
+*/
+/*
+    Returns sum of arr[0..index]. This function assumes that the array is preprocessed and partial sums of 
+    array elements are stored in BITree[]. 
+*/
+int getSum (int BITree[], int index)
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:16]: Leave a line before it

File: Fenwick_Tree/fenwick_tree.c [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+/*  n --> No. of elements present in input array. 
+    BITree[0..n] --> Array that represents Fenwick Tree(Binary Indexed Tree). 
+    arr[0..n-1] --> Input array for which prefix sum is evaluated.
+*/
+/*
+    Returns sum of arr[0..index]. This function assumes that the array is preprocessed and partial sums of 
+    array elements are stored in BITree[]. 
+*/
+int getSum (int BITree[], int index) 
+{   int sum = 0; // Iniialize result
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:17]: Code should be in next line

File: Fenwick_Tree/fenwick_tree.c [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+*/
+int getSum (int BITree[], int index) 
+{   int sum = 0; // Iniialize result 
+
+    // index in BITree[] is 1 more than the index in arr[] 
+    index = index + 1; 
+
+    // Traverse ancestors of BITree[index] 
+    while (index > 0) 
+   {    // Add current element of BITree to sum
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:24]: inconsistency in indentation

File: Fenwick_Tree/fenwick_tree.c [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    index = index + 1; 
+
+    // Traverse ancestors of BITree[index] 
+    while (index > 0) 
+   {    // Add current element of BITree to sum 
+	sum += BITree[index]; 
+
+	// Move index to parent node in getSum View 
+	index -= index & (-index); 
+    }
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:29]: inconsistency in indentation

File: Fenwick_Tree/fenwick_tree.c [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+	// Move index to parent node in getSum View 
+	index -= index & (-index); 
+    } 
+    return sum; 
+} 
+/*
+   Updates a node in Fenwick Tree at given index in BITree. The given value 'val' is added to BITree[i] and 
+   all of its ancestors in tree. 
+*/
+void updateBIT (int BITree[], int n, int index, int val)
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:36]: leave a line before it

File: Fenwick_Tree/fenwick_tree.c [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    while (index > 0) 
+   {    // Add current element of BITree to sum 
+	sum += BITree[index]; 
+
+	// Move index to parent node in getSum View 
+	index -= index & (-index); 
+    } 
+    return sum; 
+} 
+/*
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:32]: leave a line before it

File: Fenwick_Tree/fenwick_tree.c [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+   Updates a node in Fenwick Tree at given index in BITree. The given value 'val' is added to BITree[i] and 
+   all of its ancestors in tree. 
+*/
+void updateBIT (int BITree[], int n, int index, int val) 
+{   // index in BITree[] is 1 more than the index in arr[] 
+    index = index + 1; 
+
+    // Traverse all ancestors and add 'val' 
+    while (index <= n) 
+   {    // Add 'val' to current node of BI Tree
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:42]: inconsistency in indentation

File: Fenwick_Tree/fenwick_tree.c [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+{   // index in BITree[] is 1 more than the index in arr[] 
+    index = index + 1; 
+
+    // Traverse all ancestors and add 'val' 
+    while (index <= n) 
+   {    // Add 'val' to current node of BI Tree 
+        BITree[index] += val; 
+
+	// Update index to that of parent in update View 
+	index += index & (-index);
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:46]: inconsistency in indentation

File: Fenwick_Tree/fenwick_tree.c [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+    
+    //Update operation
+    freq[4] += 16;
+    updateBIT(BITree, n, 4, 16);
+    
+    //Print sum of freq[2....7] after update
+    printf("%d\n", rangeSum(2, 7, BITree));
+    return 0; 
+} 
+/*
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:91]: leave a line before it

File: Fenwick_Tree/fenwick_tree.c [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+    for (int i = 0; i < n; ++i) 
+	updateBIT(BITree , n , i , arr[i]); 
+
+    return BITree; 
+} 
+int rangeSum (int l, int r, int BITree[])
+{
+    return getSum(BITree, r) - getSum(BITree, l);
+} 
+int main()
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:69]: leave a line before it

File: Fenwick_Tree/fenwick_tree.c [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+    for (int i = 1; i <= n; ++i) 
+	BITree[i] = 0; 
+
+    // Store the actual values in BITree[] using update() 
+    for (int i = 0; i < n; ++i) 
+	updateBIT(BITree , n , i , arr[i]); 
+
+    return BITree; 
+} 
+int rangeSum (int l, int r, int BITree[])
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:65]: leave a line before it

File: Fenwick_Tree/fenwick_tree.c [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+        BITree[index] += val; 
+
+	// Update index to that of parent in update View 
+	index += index & (-index); 
+    } 
+} 
+/*
+   Constructs and returns a Fenwick Tree for given array of size n. 
+*/
+int *constructBITree (int arr[], int n)
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:52]: leave a line before it

File: Fenwick_Tree/fenwick_tree.c [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+    // Traverse all ancestors and add 'val' 
+    while (index <= n) 
+   {    // Add 'val' to current node of BI Tree 
+        BITree[index] += val; 
+
+	// Update index to that of parent in update View 
+	index += index & (-index); 
+    } 
+} 
+/*
```
Comment (somya-kapoor) [inline Fenwick_Tree/fenwick_tree.c:49]: leave a line before it

**Comment by santushtisharma10** [comment]
@somya-kapoor i have made the changes as you said,please review them

File: Fenwick_Tree/Fenwick_Tree.c [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+int getSum (int BITree[], int index) 
+{   
+    int sum = 0; // Iniialize result 
+
+    // index in BITree[] is 1 more than the index in arr[] 
+    index = index + 1; 
+
+    // Traverse ancestors of BITree[index] 
+    while (index > 0) 
+   {
```
Comment (somya-kapoor) [inline Fenwick_Tree/Fenwick_Tree.c:28]: inconsistency in indentation

File: Fenwick_Tree/Fenwick_Tree.c [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+
+    // Traverse ancestors of BITree[index] 
+    while (index > 0) 
+   {    
+	// Add current element of BITree to sum 
+	sum += BITree[index]; 
+
+	// Move index to parent node in getSum View 
+	index -= index & (-index); 
+    }
```
Comment (somya-kapoor) [inline Fenwick_Tree/Fenwick_Tree.c:34]: 4 space indentation

File: Fenwick_Tree/Fenwick_Tree.c [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+
+    // Traverse all ancestors and add 'val' 
+    while (index <= n) 
+   {    
+	// Add 'val' to current node of BI Tree 
+        BITree[index] += val; 
+
+	// Update index to that of parent in update View 
+	index += index & (-index); 
+    }
```
Comment (somya-kapoor) [inline Fenwick_Tree/Fenwick_Tree.c:56]: 4 space indentation

File: Fenwick_Tree/Fenwick_Tree.c [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+   Constructs and returns a Fenwick Tree for given array of size n. 
+*/
+
+int *constructBITree (int arr[], int n) 
+{   
+    // Create and initialize BITree[] as 0 
+    int *BITree = new int[n + 1]; 
+
+    for (int i = 1; i <= n; ++i) 
+	BITree[i] = 0;
```
Comment (somya-kapoor) [inline Fenwick_Tree/Fenwick_Tree.c:69]: 4 space indentation

File: Fenwick_Tree/Fenwick_Tree.c [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+{   
+    // Create and initialize BITree[] as 0 
+    int *BITree = new int[n + 1]; 
+
+    for (int i = 1; i <= n; ++i) 
+	BITree[i] = 0; 
+
+    // Store the actual values in BITree[] using update() 
+    for (int i = 0; i < n; ++i) 
+	updateBIT(BITree , n , i , arr[i]);
```
Comment (somya-kapoor) [inline Fenwick_Tree/Fenwick_Tree.c:73]: 4 space indentation
remove space before ,

File: Fenwick_Tree/Fenwick_Tree.c [hunk 29]
```diff
@@ -0,0 +1,118 @@
...
+} 
+
+int main() 
+{   
+    int n;
+    scanf("%d", &n);
+	
+    int freq[n];	
+    for(int i = 0; i < n; ++i)
+   {
```
Comment (somya-kapoor) [inline Fenwick_Tree/Fenwick_Tree.c:90]: inconsistency in indentation

File: Fenwick_Tree/Fenwick_Tree.c [hunk 30]
```diff
@@ -0,0 +1,118 @@
...
+    
+    //Update operation
+    freq[4] += 16;
+    updateBIT(BITree, n, 4, 16);
+    
+    //Print sum of freq[2....7] after update
+    printf("%d\n", rangeSum(2, 7, BITree));
+    return 0; 
+} 
+/*
```
Comment (somya-kapoor) [inline Fenwick_Tree/Fenwick_Tree.c:105]: Leave a line before it

Comment (santushtisharma10) [reply Fenwick_Tree/Fenwick_Tree.c:108]: @somya-kapoor Please specify the line no.

**Comment by santushtisharma10** [comment]
@somya-kapoor please review my code, i have made changes as you said

**Comment by santushtisharma10** [comment]
@somya-kapoor please review my code

**Comment by somya-kapoor** [comment]
Please squash the commits, after that it will be ready for merging

**Comment by santushtisharma10** [comment]
@somya-kapoor i have squashed my commits,please review it

File: Fenwick_Tree/Fenwick_Tree.c [hunk 31]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+/*
+   Constructs and returns a Fenwick Tree for given array of size n. 
+*/
+
+int *constructBITree(int arr[], int n)
+{
+    // Create and initialize BITree[] as 0
+    int *BITree = new int[n + 1];
```
Comment (somya-kapoor) [inline Fenwick_Tree/Fenwick_Tree.c:66]: Did you run your code, because their is no new keyword in c

**Comment by santushtisharma10** [comment]
@somya-kapoor ,i have made the correction and now its running successfully and giving the correct output as well,please review it

**Comment by santushtisharma10** [comment]
@somya-kapoor please review my code

**Pull request merged**